### DISCUSSION TITLE
A query doesnt work even though everything seems right. Can someone help point the error out

### DISCUSSION THREAD

**Discussion by Dorfeuheinz** [vinniefranco/exandra#9] [Q&A] [answered]
I have a create user query here that doesnt seem to work kindly let me know whats incorrect though nothing comes to mind immediately

```def create_user(opts) do
    keyspace = Keyword.fetch!(opts, :keyspace)
    query = """
      use #{keyspace};
      
      CREATE TYPE themes (
        id uuid PRIMARY KEY,
        name text
      );
    
      CREATE TYPE account_settings (
        visibility_public? boolean,
        theme themes
      );
      
      CREATE TABLE IF NOT EXISTS users (
        id uuid,
        email text,
        name text,
        joined_at timestamp,
        settings account_settings,
        theme themes,
        PRIMARY KEY (id, email)
      ) WITH CLUSTERING ORDER BY (joined_at DESC);
    """
    storage_toggle(:xandra_conn, query, "TABLE", "CREATED", "Already exists")
  end

  defp storage_toggle(conn, query, target, effect, error_message) do
    case Xandra.Cluster.execute(conn, query, []) do
      {:ok, %Xandra.SchemaChange{target: ^target, effect: ^effect}} -> {:ok, target <> " " <> effect}
      {:ok, %Xandra.Void{}} -> {:ok, error_message}
      {:error, %Xandra.Error{message: message}} -> message
    end
  end```
  
iex> Myapp.Scylla.create_user([keyspace: "ks"])
"line 3:2 : syntax error...\n"
It doesnt work

**Comment by vinniefranco** [comment, marked as answer]
Hey there! Thanks for the interest, and sorry there are no docs to get you started. 

This is an Ecto adapter that wraps Xandra into the Etco DSL that we all love by following the custom adapter section detailed here https://hexdocs.pm/ecto_sql/Ecto.Adapters.SQL.html#module-custom-adapters. If you want to use Xandra directly, then I recommend doing so.

If you do want to use this adapter - then you'll setup it more like postgres/mysql for Ecto. For example:

```elixir
# config.exs

config :myall, Myapp.Repo
  default_consistency: :one,
  keyspace: "myapp",
  log_level: :debug,
  primary_key: [name: :id, type: :binary_id],
  nodes: ["myscyllanodeaddr"]
  protocol_version: :v4
```

```elixir
# repo.ex

defmodule Myapp.Repo do
  use Ecto.Repo, otp_app: :myapp, adapter: Exandra
end
```

Then define your schemas (example: https://github.com/vinniefranco/exandra/blob/main/test/exandra_test.exs#L4-L41)

Finally, use the normal ecto query approach https://hexdocs.pm/ecto/Ecto.Query.html

Of coure, I'm handwaving migrations and keyspace creation at this point because those are very much a work in progress. Just like the rest of the library :)

Hopefully, this gets you pointed in the right direction!

Cheers,
V

**Reply by Dorfeuheinz** [reply]
Oh thankyou so much for the reply. tho im sorry, I forgot to update my comment from earlier. I have fixed it & now everything works just fine thanks. Tho yeah, im going no adapters with this 1, since working with drivers are kindoff necessary at this point, tho I too follow your work which is IMHO incredible. It really helped a lot. Ill keep following the progress, since I too need help with seeing how to write migrations which is the last step. Thanks for the reply

**Reply by Dorfeuheinz** [reply]
I think my question was removed or something, so ill ask again. How are you planning on using migrations? My proposed solution that came to my mind was to use default ecto migrations & write macro to use that as a means to generate corresponding cassandra queries that can be timestamped etc. so as to build the db via migration files. How are you doing it & kindly let me know if my solution is acceptable or if you have a different plan.

**Reply by vinniefranco** [reply]
@Dorfeuheinz thank you for the kind words! The good news is that @whatyouhide is in here and really cleaning things up for the next release. Which I'm pleased to report will have full migrations.

**Reply by Dorfeuheinz** [reply]
oh great thanks so much. That will help.

**Reply by vinniefranco** [reply]
https://hexdocs.pm/exandra/0.2.0/Exandra.html released!
